src/proxy/outbound.rs

```diff
@@ -202,10 +204,58 @@ impl OutboundConnection {
                 self.proxy_to_tcp(source_stream, &req, &result_tracker)
                     .await
             }
+            Protocol::DOUBLEHBONE => {
+                self.proxy_to_double_hbone(source_stream, source_addr, &req, &result_tracker)
+                    .await
+            }
         };
         result_tracker.record(res)
     }
 
+    async fn proxy_to_double_hbone(
+        &mut self,
+        stream: TcpStream,
+        remote_addr: SocketAddr,
+        req: &Request,
+        connection_stats: &ConnectionResult,
+    ) -> Result<(), Error> {
+        // Outer HBONE
+        let upgraded = Box::pin(self.send_hbone_request(remote_addr, req)).await?;
+
+        // Inner HBONE
+        let upgraded = TokioH2Stream::new(upgraded);
+        let inner_workload = pool::WorkloadKey {
+            src_id: req.source.identity(),
+            dst_id: req.upstream_sans.clone(),
+            src: remote_addr.ip(),
+            dst: req.actual_destination,
+        };
+        let request = self.create_hbone_request(remote_addr, req);
+
+        let (_conn_client, inner_upgraded, drain_tx, driver_task) = self
+            .pool
+            .send_request_unpooled(upgraded, &inner_workload, request)
+            .await?;
+        let inner_upgraded = inner_upgraded?;
+        let res = copy::copy_bidirectional(
+            copy::TcpStreamSplitter(stream),
+            inner_upgraded,
+            connection_stats,
+        )
+        .await;
+
+        // This always drops ungracefully
+        // drop(conn_client);
+        // tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+        // drain_tx.send(true).unwrap();
+        // tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+        drain_tx.send(true).unwrap();
+        let _ = driver_task.await;
+        // this sleep is important, so we have a race condition somewhere
+        // tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+        res
```

Comment:
Does anybody have any info on how to properly drop/terminate H2 connections over stream with nontrivial drops (e.g. shutting down TLS over HTTP2 CONNECT). Right now, I'm just dropping things/aborting tasks randomly until something works

Reply:
Are you asking about how to cleanup after, for example, a RST_STREAM to the inner tunnel? Or something else

Reply:
Kinda. I mostly mean the outer TLS stream because that's what I've looked at. It seems like if I drop `conn_client` before termination `driver_task` the TCP connection will close **without** sending close notifies. So yes, I'm asking if there is a way to explicitly do cleanup rather than relying on implicit drops.

Reply:
I see the code changed; do you still need help figuring this out?

Reply:
Im still not confident in it. It works (on my machine), but I couldn't find any docs on proper connection termination/dropping.

Reply:
Seems like we ignore shutdown errors so I'm not going to worry about this